service/history/shard/coordinator.go

```diff
@@ -0,0 +1,214 @@
+	}
+}
+
+func (c *coordinatorImpl) Start() {
+
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	go c.receiveFailoverMarkersLoop()
+	go c.sendFailoverMarkerLoop()
+}
+
+func (c *coordinatorImpl) Stop() {
+
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(c.shutdownChan)
+}
+
+func (c *coordinatorImpl) SendFailoverMarkers(
+	shardID int,
+	markers []*replicator.FailoverMarkerAttributes,
+) <-chan error {
+
+	respCh := make(chan error, 1)
+	c.sendChan <- &request{
+		shardID: shardID,
+		markers: markers,
+		respCh:  respCh,
+	}
+
+	return respCh
+}
+
+func (c *coordinatorImpl) ReceiveFailoverMarkers(
+	shardID int,
+	markers []*replicator.FailoverMarkerAttributes,
+) {
+
+	c.receiveChan <- &request{
+		shardID: shardID,
+		markers: markers,
+	}
+}
+
+func (c *coordinatorImpl) receiveFailoverMarkersLoop() {
+
+	for {
+		select {
+		case <-c.shutdownChan:
+			return
+		case request := <-c.receiveChan:
+			for _, marker := range request.markers {
+				domainID := marker.GetDomainID()
+
+				if record, ok := c.recorder[domainID]; ok {
+					// if the local failover version is smaller than the new received marker,
+					// it means there is another failover happened and the local one should be invalid.
+					if record.failoverVersion < marker.GetFailoverVersion() {
+						delete(c.recorder, domainID)
+					}
+				}
+
+				if _, ok := c.recorder[domainID]; !ok {
+					// initialize the failover record
+					c.recorder[marker.GetDomainID()] = &failoverRecord{
+						failoverVersion: marker.GetFailoverVersion(),
+						shards:          make(map[int]struct{}),
+					}
+					return
+				}
+
+				record := c.recorder[domainID]
+				record.shards[request.shardID] = struct{}{}
+				if len(record.shards) == c.config.NumberOfShards {
+					// TODO: update domain from pend_active to active
+					delete(c.recorder, domainID)
+				}
+			}
+		}
+	}
+}
+
+func (c *coordinatorImpl) sendFailoverMarkerLoop() {
+
+	timer := time.NewTimer(backoff.JitDuration(
+		c.config.FailoverMarkerHeartbeatInterval(),
+		c.config.FailoverMarkerHeartbeatTimerJitterCoefficient(),
+	))
+
+	requestByShard := make(map[int]*request)
+
+	for {
+		select {
+		case <-c.shutdownChan:
+			timer.Stop()
+			return
+		case request := <-c.sendChan:
+			// Here we only add the request to map. We will wait until timer fires to send the request to remote.
+			if req, ok := requestByShard[request.shardID]; ok && req != request {
+				// during shard movement, duplicated requests can appear
+				// if shard moved from this host, to this host.
+				c.logger.Warn("Failover marker heartbeat request already exist for shard.")
+				if req.respCh != nil {
+					close(req.respCh)
+				}
+			}
+			requestByShard[request.shardID] = request
+		case <-timer.C:
+			if len(requestByShard) > 0 {
+				var err error
+				//TODO: Remote calls to send failover markers
```

Comment:
the markers should be part of replication queue (plus making sure this host own the shard)?
if so why queueing here?

Reply:
The queueing will be done inside the failover callback. here is each heart heartbeat the markers to the coordinator.